src/main/java/io/jenkins/plugins/opentelemetry/backend/elastic/ElasticsearchLogStorageScrollingRetriever.java

```diff
@@ -0,0 +1,224 @@
+            throw new IllegalArgumentException("Elasticsearch Index Pattern cannot be blank");
+        }
+        this.indexPattern = indexPattern;
+
+        BasicCredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+        credentialsProvider.setCredentials(AuthScope.ANY, elasticsearchCredentials);
+
+        RestClient restClient = RestClient.builder(HttpHost.create(elasticsearchUrl))
+            .setHttpClientConfigCallback(httpClientBuilder -> httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider))
+            .build();
+        RestClientTransport elasticsearchTransport = new RestClientTransport(restClient, new JacksonJsonpMapper());
+        this.elasticsearchClient = new ElasticsearchClient(elasticsearchTransport);
+    }
+
+    @Nonnull
+    @Override
+    public LogsQueryResult overallLog(@Nonnull String traceId, @Nonnull String spanId, @Nullable LogsQueryContext logsQueryContext) throws IOException {
+        // https://www.elastic.co/guide/en/elasticsearch/reference/7.17/scroll-api.html
+
+        Charset charset = StandardCharsets.UTF_8;
+        boolean complete;
+        String newScrollId;
+        List<Hit<ObjectNode>> hits;
+        if (logsQueryContext == null) {
+            SearchRequest searchRequest = new SearchRequest.Builder()
+                .scroll(builder -> builder.time("30s"))
+                .size(PAGE_SIZE)
+                .sort(sortBuilder -> sortBuilder.field(fieldBuilder -> fieldBuilder.field(TIMESTAMP).order(SortOrder.Asc)))
+                .query(queryBuilder ->
+                    queryBuilder.match(
+                        matchQueryBuilder -> matchQueryBuilder.field("trace.id").query(
+                            fieldValueBuilder -> fieldValueBuilder.stringValue(traceId))))
+                // .fields() TODO narrow down the list fields to retrieve - we probably have to look at a source filter
+                .build();
+            logger.log(Level.INFO, "Retrieve logs for traceId: " + traceId);
+            SearchResponse<ObjectNode> searchResponse = this.elasticsearchClient.search(searchRequest, ObjectNode.class);
+            hits = searchResponse.hits().hits();
+            newScrollId = searchResponse.scrollId();
+        } else {
+            ElasticsearchLogsQueryScrollingContext context = (ElasticsearchLogsQueryScrollingContext) logsQueryContext;
+            logger.log(Level.INFO, "Retrieve logs with scrollId: " + context.scrollId + " for traceId: " + traceId);
+            ScrollRequest scrollRequest = ScrollRequest.of(builder -> builder.scrollId(context.scrollId));
+            ScrollResponse<ObjectNode> scrollResponse = this.elasticsearchClient.scroll(scrollRequest, ObjectNode.class);
+            hits = scrollResponse.hits().hits();
+            newScrollId = context.scrollId; // TODO why doesn't the scroll response hold a new scrollId? scrollResponse.scrollId();
+        }
+        complete = hits.size() != PAGE_SIZE; // TODO is there smarter?
```

Comment:
@swallez what's the best way to understand if the search request returned all the entries or if I should scroll more? as `scrollResponse.scrollId` is always `null`, I end up doing one wasted query if the number of entries of the response is a multiple of `PAGE_SIZE`